core/src/main/scala/kafka/coordinator/group/MemberMetadata.scala

```diff
@@ -85,16 +85,16 @@ private[group] class MemberMetadata(var memberId: String,
     }
   }
 
-  def shouldKeepAlive(deadlineMs: Long): Boolean = {
+  def hasSatisfiedHeartbeat: Boolean = {
     if (isNew) {
-      // New members are expired after the static join timeout
-      latestHeartbeat + GroupCoordinator.NewMemberJoinTimeoutMs > deadlineMs
+      // New members can be expired while awaiting join, so we have to check this first
```

Comment:
Hmm, do we still need to rely on `isNew` here? Could the condition just be `heartbeatSatisfied || isAwaitingJoin || isAwaitingSync`?

Reply:
The awkward thing about the `isNew` case is that we want the member to fail while `isAwaitingJoin` is true. This is attempting to handle the case for old versions of the protocol where a client sends JoinGroup with empty memberId and then times out. I was debating how important this logic is, but I seem to recall a case in the past where the group grew very large because new members kept timing out and retrying.

Reply:
Fair enough; originally I thought we can now just remove this `isNew` flag now but I think I buy your arguments now.